packages/client-engine-runtime/src/interpreter/renderQuery.ts

```diff
@@ -15,16 +27,62 @@ export function renderQuery({ query, params }: QueryPlanDbQuery, scope: ScopeBin
       throw new Error(`Missing value for query variable ${param.prisma__value.name}`)
     }
 
-    return value
+    return value as PrismaValue
   })
+}
+
+function renderTemplateSql(
+  fragments: Fragment[],
+  placeholderFormat: PlaceholderFormat,
+  params: PrismaValue[],
+): SqlQuery {
+  let paramIndex = 0
+  let placeholderNumber = 1
+  const flattenedParams: PrismaValue[] = []
+  const sql = fragments
+    .map((fragment) => {
+      switch (fragment.type) {
+        case 'parameter':
+          flattenedParams.push(params[paramIndex++])
+          return formatPlaceholder(placeholderFormat, placeholderNumber++)
+
+        case 'stringChunk':
+          return fragment.value
+
+        case 'parameterTuple': {
+          const paramValue = params[paramIndex++]
+          const paramArray = Array.isArray(paramValue) ? paramValue : [paramValue]
+          const placeholders =
+            paramArray.length == 0
+              ? 'NULL'
+              : paramArray
+                  .map((value) => {
+                    flattenedParams.push(value)
+                    return formatPlaceholder(placeholderFormat, placeholderNumber++)
+                  })
+                  .join(',')
+          return `(${placeholders})`
+        }
+
+        default:
+          return fragment // never
+      }
+    })
+    .join('')
+
+  return renderRawSql(sql, flattenedParams)
+}
 
-  const { query: renderedQuery, params: expandedParams } = renderQueryTemplate({ query, params: substitutedParams })
+function formatPlaceholder(placeholderFormat: PlaceholderFormat, placeholderNumber: number): string {
+  return placeholderFormat.hasNumbering ? `${placeholderFormat.prefix}${placeholderNumber}` : placeholderFormat.prefix
+}
 
-  const argTypes = expandedParams.map((param) => toArgType(param as PrismaValue))
+function renderRawSql(sql: string, params: PrismaValue[]): SqlQuery {
+  const argTypes = params.map((param) => toArgType(param as PrismaValue))
```

Comment:
why is `as PrismaValue` necessary here? `params` is already `PrismaValue[]`

Reply:
Nice catch. It is interesting that the unnecessary cast wasn't cough by the code inspection here.